lib/linked_list.rb

```diff
@@ -12,152 +12,303 @@ def initialize(value, next_node = nil)
+      return nil
     end
+  end
 
-    # checks if the linked list has a cycle. A cycle exists if any node in the
-    # linked list links to a node already visited.
-    # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
-    # Space Complexity
-    def has_cycle
-      raise NotImplementedError
+  # find the nth node from the end and return its value
+  # assume indexing starts at 0 while counting to n
+  # Time Complexity: 0(2n) --> O(n)
+  # Space Complexity: O(1)
+  def find_nth_from_end(n)
+    if n == 0
+      return self.get_last
+    elsif n == self.length - 1
+      return self.get_first
+    else
+      index = (self.length - 1) - n
+      if index > -1
+        return self.get_at_index(index)
+      else
+        return nil
+      end
     end
+  end
 
+  # checks if the linked list has a cycle. A cycle exists if any node in the
+  # linked list links to a node already visited.
+  # returns true if a cycle is found, false otherwise.
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def has_cycle
+    move_by_one = @head
+    move_by_two = @head
 
-    # Additional Exercises 
-    # returns the value in the first node
-    # returns nil if the list is empty
-    # Time Complexity:  
-    # Space Complexity
-    def get_first
-      raise NotImplementedError
+    until move_by_two == nil && move_by_two.next == nil
+      move_by_one = move_by_one.next
+      move_by_two = move_by_two.next.next
+      return true if move_by_one == move_by_two
     end
+    return false
+  end
 
-    # method that inserts a given value as a new last node in the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_last(value)
-      raise NotImplementedError
+  # Additional Exercises
+  # returns the value in the first node
+  # returns nil if the list is empty
+  # Time Complexity: O(1)
+  # Space Complexity: O(1)
+  def get_first
+    if @head == nil
+      return nil
+    else 
+      return @head.data
     end
+  end
 
-    # method that returns the value of the last node in the linked list
-    # returns nil if the linked list is empty
-    # Time Complexity:  
-    # Space Complexity
-    def get_last
-      raise NotImplementedError
+  # method that inserts a given value as a new last node in the linked list
+  # Time Complexity: O(n)
+  # Space Complexity: 
+  def add_last(value)
+    if @head
+      current_node = @head
+      until current_node.next == nil
+        current_node = current_node.next
+      end
+      
+      current_node.next = Node.new(value)
+    else
+      @head = Node.new(value)
     end
+  end
 
-    # method to insert a new node with specific data value, assuming the linked
-    # list is sorted in ascending order
-    # Time Complexity:  
-    # Space Complexity
-    def insert_ascending(value)
-      raise NotImplementedError
+  # method that returns the value of the last node in the linked list
+  # returns nil if the linked list is empty
+  # Time Complexity:

```

Comment:
I think you understand the idea here, but I'm just noting the absence one last time.